rllib/models/specs/checker.py

```diff
@@ -0,0 +1,290 @@
+@DeveloperAPI(stability="alpha")
+def check_input_specs(
+    input_spec: str,
+    *,
+    filter: bool = False,
+    cache: bool = False,
+):
+    """A general-purpose spec checker decorator for Neural Network base classes.
+
+    This is a stateful decorator
+    (https://realpython.com/primer-on-python-decorators/#stateful-decorators) to
+    enforce input specs for any instance method that has an argument named
+    `input_data` in its args.
+
+    It also allows you to filter the input data dictionary to only include those keys
+    that are specified in the model specs. It also allows you to cache the validation
+    to make sure the spec is only validated once in the entire lifetime of the instance.
+
+    Examples (See more exmaples in ../tests/test_specs_dict.py):
+
+        >>> class MyModel(nn.Module):
+        ...     @property
+        ...     def input_spec(self):
+        ...         return {"obs": TensorSpec("b, d", d=64)}
+        ...
+        ...     @check_input_specs("input_spec")
+        ...     def forward(self, input_data, return_loss=False):
+        ...         ...
+
+        >>> model = MyModel()
+        >>> model.forward({"obs": torch.randn(32, 64)}) # No error
+        >>> model.forward({"obs": torch.randn(32, 32)}) # raises ValueError
+
+    Args:
+        func: The instance method to decorate. It should be a callable that takes
+            `self` as the first argument, `input_data` as the second argument and any
+            other keyword argument thereafter.
+        input_spec: `self` should have an instance attribute whose name matches the
+            string in input_spec and returns the `SpecDict`, `Spec`, or simply the
+            `Type` that the `input_data` should comply with. It can alos be None or
+            empty list / dict to enforce no input spec.
+        filter: If True, and `input_data` is a nested dict the `input_data` will be
+            filtered by its corresponding spec tree structure and then passed into the
+            implemented function to make sure user is not confounded with unnecessary
+            data.
+        cache: If True, only checks the data validation for the first time the
+            instance method is called.
+
+    Returns:
+        A wrapped instance method. In case of `cache=True`, after the first invokation
+        of the decorated method, the intance will have `__checked_input_specs_cache__`
+        attribute that stores which method has been invoked at least once. This is a
+        special attribute that can be used for the cache itself. The wrapped class
+        method also has a special attribute `__checked_input_specs__` that marks the
+        method as decorated.
+    """
+
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(self, input_data, **kwargs):
+            if cache and not hasattr(self, "__checked_input_specs_cache__"):
+                self.__checked_input_specs_cache__ = {}
+
+            input_data_ = input_data
```

Comment:
Can we use another name for `input_data_` for more clarity? E.g. "processed_input_data"?

Reply:
fixed.